File: reports/general/dupefilenames.py
# Public domain; MZMcBride, Tim Landscheidt; 2011, 2013

"""
Report class for largely duplicative file names
"""

import reports

class report(reports.report):
    def get_title(self):
        return 'Largely duplicative file names'

    def get_preamble_template(self):
        return '''Largely duplicative file names (limited to the first 1000 entries); \
data as of <onlyinclude>%s</onlyinclude>.'''

    def get_table_columns(self):
        return ['Normalized name', 'Count', 'Real names']

    def get_table_rows(self, conn):
        cursor = conn.cursor()
        cursor.execute('''
        /* dupefilenames.py SLOW_OK */
        SELECT
          LOWER(CONVERT(page_title USING utf8)),
          GROUP_CONCAT(CONVERT(page_title USING utf8) SEPARATOR '|'),
          COUNT(*)
        FROM page
        WHERE page_namespace = 6
        AND page_is_redirect = 0
        GROUP BY 1
        HAVING COUNT(*) > 1
        LIMIT 1000;
        ''')
        for norm_name, orig_names_str, count in cursor:
            orig_names = []
            for name in orig_names_str.split('|'):
                name = u'[[:File:%s|%s]]' % (name, name)
                orig_names.append(name)
            orig_name = ', '.join(orig_names)
            yield [norm_name, str(count), orig_name]

        cursor.close()

File: dbreps2/src/general/dupefilenames.rs
// released under public domain; MZMcBride, Tim Landscheidt, Deadbeef; 2011, 2013, 2023

use anyhow::Result;
use dbreps2::{Frequency, Report};
use mysql_async::prelude::Queryable;
use mysql_async::Conn;

pub struct DupeFileNames;

pub struct Row {
    norm_name: String,
    count: usize,
    orig_names_str: String,
}

impl Report<Row> for DupeFileNames {
    fn title(&self) -> &'static str {
        "Largely duplicative file names"
    }

    fn intro(&self) -> &'static str {
        "Largely duplicative file names (limited to the first 1000 entries)"
    }

    fn headings(&self) -> Vec<&'static str> {
        vec!["Normalized name", "Count", "Real names"]
    }

    fn frequency(&self) -> Frequency {
        Frequency::Daily
    }

    fn query(&self) -> &'static str {
        "
        /* dupefilenames.py SLOW_OK */
        SELECT
          LOWER(CONVERT(page_title USING utf8mb4)),
          GROUP_CONCAT(CONVERT(page_title USING utf8mb4) SEPARATOR '|'),
          COUNT(*)
        FROM page
        WHERE page_namespace = 6
        AND page_is_redirect = 0
        GROUP BY 1
        HAVING COUNT(*) > 1
        LIMIT 1000;
        "
    }

    async fn run_query(&self, conn: &mut Conn) -> Result<Vec<Row>> {
        Ok(conn
            .query_map(self.query(), |(norm_name, orig_names_str, count)| Row {
                norm_name,
                count,
                orig_names_str,
            })
            .await?)
    }

    fn format_row(&self, row: &Row) -> Vec<String> {
        vec![
            row.norm_name.clone(),
            row.count.to_string(),
            row.orig_names_str
                .split('|')
                .map(|x| format!("[[:File:{x}|{x}]]"))
                .collect::<Vec<_>>()
                .join(", "),
        ]
    }

    fn code(&self) -> &'static str {
        include_str!("dupefilenames.rs")
    }
}
